File: src/libs/rust/base/src/env.rs
/*
 * Copyright (C) 2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

//! Provides access to the program environment

use core::iter;
use core::ops::FnOnce;

use derivative::Derivative;

use num_enum::{IntoPrimitive, TryFromPrimitive};

use crate::boxed::Box;
use crate::cell::LazyStaticRefCell;
use crate::cfg;
use crate::col::{String, ToString, Vec};
use crate::format;
use crate::mem::VirtAddr;
use crate::util;

#[derive(Copy, Clone, Debug, Default, Eq, PartialEq, IntoPrimitive, TryFromPrimitive)]
#[repr(u64)]
pub enum Platform {
    #[default]
    Gem5,
    Hw,
}

#[derive(Copy, Clone, Derivative, Debug)]
#[derivative(Default)]
#[repr(C)]
pub struct BootEnv {
    pub platform: Platform,
    pub tile_id: u64,
    pub tile_desc: u64,
    pub argc: u64,
    pub argv: u64,
    pub envp: u64,
    pub kenv: u64,
    pub raw_tile_count: u64,
    #[derivative(Default(value = "[0u64; cfg::MAX_TILES * cfg::MAX_CHIPS]"))]
    pub raw_tile_ids: [u64; cfg::MAX_TILES * cfg::MAX_CHIPS],
}

#[derive(Copy, Clone, Debug, Default)]
#[repr(C)]
pub struct BaseEnv {
    pub boot: BootEnv,

    pub shared: u64,
    pub sp: u64,
    pub entry: u64,
    pub closure: u64,
    pub heap_size: u64,
    pub first_std_ep: u64,
    pub first_sel: u64,
    pub act_id: u64,

    pub rmng_sel: u64,
    pub pager_sess: u64,
    pub pager_sgate: u64,

    pub mounts_addr: u64,
    pub mounts_len: u64,

    pub fds_addr: u64,
    pub fds_len: u64,

    pub data_addr: u64,
    pub data_len: u64,
}

/// Collects the strings and pointers for the given slice of arguments to pass to a program.
///
/// The strings and pointers are stored relative to the given address (`addr`).
///
/// Returns a tuple of the strings, pointers, and the final address (for a another call of
/// `collect_args`).
pub fn collect_args<S>(args: &[S], addr: VirtAddr) -> (Vec<u8>, Vec<VirtAddr>, VirtAddr)
where
    S: AsRef<str>,
{
    let mut arg_ptr = Vec::<VirtAddr>::new();
    let mut arg_buf = Vec::new();

    let mut arg_addr = addr;
    for s in args {
        // push argv entry
        arg_ptr.push(arg_addr);

        // push string
        let arg = s.as_ref().as_bytes();
        arg_buf.extend_from_slice(arg);

        // 0-terminate it
        arg_buf.push(b'\0');

        arg_addr += arg.len() + 1;
    }
    arg_ptr.push(VirtAddr::null());

    (arg_buf, arg_ptr, arg_addr)
}

pub fn boot() -> &'static BootEnv {
    // safety: the cast is okay because we trust our loader to put the environment at that place
    unsafe { &*(cfg::ENV_START.as_ptr()) }
}

/// The closure used by `Activity::run`
pub struct Closure {
    func: Option<Box<dyn FnOnce() -> i32 + Send>>,
}

impl Closure {
    /// Creates a new object for given closure
    pub fn new<F>(func: Box<F>) -> Self
    where
        F: FnOnce() -> i32 + Send + 'static,
    {
        Closure { func: Some(func) }
    }

    /// Calls the closure (can only be done once) and returns its exit code
    pub fn call(&mut self) -> i32 {
        match self.func.take() {
            Some(c) => c(),
            None => 1,
        }
    }
}

/// The command line argument iterator
///
/// # Examples
///
/// ```
/// for arg in env::args() {
///     println!("{}", arg);
/// }
/// ```
#[derive(Copy, Clone)]
pub struct Args {
    pos: isize,
}

impl Args {
    fn arg(self, idx: isize) -> &'static str {
        // safety: we assume that our loader has put valid strings at argv
        unsafe {
            let args = boot().argv as *const u64;
            let arg = *args.offset(idx);
            util::cstr_to_str(arg as *const i8)
        }
    }

    /// Returns true if there are no arguments
    pub fn is_empty(self) -> bool {
        self.len() == 0
    }

    /// Returns the number of arguments
    pub fn len(self) -> usize {
        boot().argc as usize
    }
}

impl iter::Iterator for Args {
    type Item = &'static str;

    fn next(&mut self) -> Option<Self::Item> {
        if self.pos < boot().argc as isize {
            let arg = self.arg(self.pos);
            self.pos += 1;
            Some(arg)
        }
        else {
            None
        }
    }
}

/// Returns the argument iterator
pub fn args() -> Args {
    Args { pos: 0 }
}

/// The environment-variable iterator
///
/// # Examples
///
/// ```
/// for (key, val) in env::vars() {
///     println!("{}={}", key, val);
/// }
/// ```
#[derive(Default)]
pub struct Vars {
    pos: isize,
}

impl iter::Iterator for Vars {
    type Item = &'static str;

    fn next(&mut self) -> Option<Self::Item> {
        let args = boot().envp as *const u64;
        if args.is_null() {
            return None;
        }

        // safety: we assume that our loader has put valid pointers and strings in envp
        let arg = unsafe { *args.offset(self.pos) } as *const i8;
        if !arg.is_null() {
            unsafe {
                let var = util::cstr_to_str(arg);
                self.pos += 1;
                Some(var)
            }
        }
        else {
            None
        }
    }
}

static VARS: LazyStaticRefCell<Vec<String>> = LazyStaticRefCell::default();

/// Returns the value of the environment variable with given key.
pub fn var<K: AsRef<str>>(key: K) -> Option<String> {
    // try to find the value from an iterator of strings of <key>=<value>
    fn find_value<'s>(it: impl iter::Iterator<Item = &'s str>, key: &str) -> Option<String> {
        it.map(|p| {
            let mut pair = p.splitn(2, '=');
            (pair.next().unwrap(), pair.next().unwrap())
        })
        .find(|(k, _v)| *k == key)
        .map(|(_k, v)| v.to_string())
    }

    // if we have already copied the env vars, use the copy
    if VARS.is_some() {
        find_value(VARS.borrow().iter().map(|s| &s[..]), key.as_ref())
    }
    else {
        find_value(Vars::default(), key.as_ref())
    }
}

/// Returns the environment-variable iterator, containing each variable as a pair of key and value.
pub fn vars() -> Vec<(String, String)> {
    // split string into pair of key and value
    fn to_pair<S: AsRef<str>>(p: S) -> (String, String) {
        let mut pair = p.as_ref().splitn(2, '=');
        (
            pair.next().unwrap().to_string(),
            pair.next().unwrap().to_string(),
        )
    }

    // if we have already copied the env vars, use the copy
    if VARS.is_some() {
        VARS.borrow().iter().map(to_pair).collect()
    }
    else {
        Vars::default().map(to_pair).collect()
    }
}

/// Returns the environment-variable iterator, containing each variable as a single string in the
/// form of `key`=`value`.
pub fn vars_raw() -> Vec<String> {
    if VARS.is_some() {
        VARS.borrow().iter().map(|p| p.to_string()).collect()
    }
    else {
        Vars::default().map(|p| p.to_string()).collect()
    }
}

/// Sets the environment variable with given key to given value.
pub fn set_var<K: AsRef<str>, V: AsRef<str>>(key: K, val: V) {
    assert!(!key.as_ref().contains('='));

    // adding/changing a variable always requires a copy
    if !VARS.is_some() {
        VARS.set(Vars::default().map(|s| s.to_string()).collect());
    }

    // therefore we can forget about `Vars::new()` here.
    let mut var_vec = VARS.borrow_mut();
    // is there a variable with that name (string starts with key and the '=' follows directly)?
    if let Some(pair) = var_vec.iter_mut().find(|p| {
        p.starts_with(key.as_ref())
            && p.chars().position(|c| c == '=').unwrap() == key.as_ref().len()
    }) {
        // then just change the value
        *pair = format!("{}={}", key.as_ref(), val.as_ref());
    }
    else {
        var_vec.push(format!("{}={}", key.as_ref(), val.as_ref()));
    }
}

/// Removes the environment variable with given key.
pub fn remove_var<K: AsRef<str>>(key: K) {
    assert!(!key.as_ref().contains('='));

    // removing a variable always requires a copy
    if !VARS.is_some() {
        VARS.set(Vars::default().map(|s| s.to_string()).collect());
    }

    let mut var_vec = VARS.borrow_mut();
    // keep all variables with different keys
    var_vec.retain(|p| {
        !p.starts_with(key.as_ref())
            || p.chars().position(|c| c == '=').unwrap() != key.as_ref().len()
    });
}

File: src/libs/base/arch/kachel/Env.cc
/*
 * Copyright (C) 2015-2018 Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * Copyright (C) 2019-2022 Nils Asmussen, Barkhausen Institut
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Common.h>
#include <base/stream/Serial.h>
#include <base/CPU.h>
#include <base/Env.h>

#include <exception>
#include <functional>
#include <stdlib.h>

typedef void (*constr_func)();

extern constr_func CTORS_BEGIN;
extern constr_func CTORS_END;

EXTERN_C void __m3_init_libc(int argc, char **argv, char **envp);
EXTERN_C void __cxa_finalize(void *);
EXTERN_C void _init();
EXTERN_C int main(int argc, char **argv);

namespace m3 {

OStream &operator<<(OStream &os, const Env &senv) {
    os << "tile_id        : " << senv.tile_id << "\n";
    os << "tile_desc      : " << fmt(senv.tile_desc, "#x") << "\n";
    os << "argc         : " << senv.argc << "\n";
    os << "argv         : " << fmt(senv.argv, "p") << "\n";
    os << "heap_size    : " << fmt(senv.heap_size, "#x") << "\n";
    os << "sp           : " << fmt(senv.sp, "p") << "\n";
    os << "entry        : " << fmt(senv.entry, "p") << "\n";
    os << "shared       : " << senv.shared << "\n";
    os << "first_std_ep : " << senv.first_std_ep << "\n";
    os << "first_sel    : " << senv.first_sel << "\n";
    os << "act_id       : " << senv.act_id << "\n";
    os << "lambda       : " << fmt(senv.lambda, "p") << "\n";
    os << "rmng_sel     : " << senv.rmng_sel << "\n";
    os << "pager_sess   : " << senv.pager_sess << "\n";
    os << "mounts_addr  : " << fmt(senv.mounts_addr, "p") << "\n";
    os << "mounts_len   : " << senv.mounts_len << "\n";
    os << "fds_addr     : " << senv.fds_addr << "\n";
    os << "fds_len      : " << fmt(senv.fds_len, "p") << "\n";
    os << "data_addr    : " << senv.data_addr << "\n";
    os << "data_len     : " << fmt(senv.data_len, "p") << "\n";
    return os;
}

void Env::call_constr() {
    _init();
    for(constr_func *func = &CTORS_BEGIN; func < &CTORS_END; ++func)
        (*func)();
}

static char **rewrite_args(uint64_t *args, int count) {
    char **nargs = new char*[count + 1];
    for(int i = 0; i < count; ++i)
        nargs[i] = reinterpret_cast<char*>(args[i]);
    nargs[count] = nullptr;
    return nargs;
}

void Env::run() {
    Env *e = env();

    // ensure that the heap is initialized before potentially cloning argv below
    m3::Heap::init();

    int argc = static_cast<int>(e->argc);
    char **argv = reinterpret_cast<char**>(e->argv);
    char **envp = reinterpret_cast<char**>(e->envp);
    if(sizeof(char*) != sizeof(uint64_t)) {
        uint64_t *envp64 = reinterpret_cast<uint64_t*>(e->envp);
        int envcnt = 0;
        for(; envp64 && *envp64; envcnt++)
            envp64++;
        envp = rewrite_args(reinterpret_cast<uint64_t*>(e->envp), envcnt);
        argv = rewrite_args(reinterpret_cast<uint64_t*>(e->argv), argc);
    }

    __m3_init_libc(argc, argv, envp);
    Env::init();

    int res;
    if(e->lambda) {
        auto func = reinterpret_cast<int(*)()>(e->lambda);
        res = (*func)();
    }
    else
        res = main(argc, argv);

    ::exit(res);
    UNREACHED;
}

USED void Env::exit(int code, bool abort) {
    if(!abort)
        __cxa_finalize(nullptr);
    backend()->exit(code);
    UNREACHED;
}

}
